igraph_is_graphical and igraph_is_bigraphical

szhorvat opened the pull request:
Closes #1432 .

This draft is open for general comments on the API/design/documentation. Feedback is welcome.

I still need to verify the implementation, add tests, and possibly complete `igraph_i_is_graphical_undirected_loopy_simple` if all goes well. 

I chose the `SW` suffix constants that represent bitwise switches and can be combined with `|`.

szhorvat commented:
The intent is to deprecate `igraph_is_degree_sequence` and `igraph_is_graphical_degree_sequence` in favour of `igraph_is_graphical`.  I opted to leave the implementations of `igraph_is_degree_sequence` and `igraph_is_graphical_degree_sequence` as they are (even though a bit of code is duplicated) because there are some tiny differences: if the in- and out-degree sequence were not of the same length, previously `false` was returned. Now it throws an error for this case.

szhorvat commented:
I should note that I plan to extend `igraph_realize_degree_sequence` in the same way, but that will be a followup PR.

szhorvat commented:
@ntamas No objections so far with the bit-switches (even though one of the constants I introduced set 2 bits at once)? :-)

ntamas commented:
Seems to be okay so far! I haven't checked the actual implementation steps yet, only the API. I'll go through the implementation once I'm back from holidays (Thursday-ish).

szhorvat commented:
> I haven't checked the actual implementation steps yet,

Don't do that yet, I am still making changes. I only wanted feedback on the API. Thanks!

szhorvat commented:
I gave up hope on keeping commits organized. We are going to do squash-and-merge. Since the last two commits are all mixed up and only make sense together, please only look at the final result when reviewing.

ntamas commented:
Will do, thanks! (That's my default mode of operation anyway).

szhorvat commented:
A summary of what's in this PR:

 - Deprecate `is_graphical_degree_sequence` and `is_degree_sequence` in favour of `is_graphical`.
 - `is_graphical` now works for directed/undirected graphs, loops/no-loops, multi-edges/simple-edges. That's `2*2*2 = 8` cases.
 - The old `is_degree_sequence` corresponded to loops/multi-edges.
 - Undirected simple graph case gets a new implementation (Cloteaux), which is both faster and O(n) (not O(n log n)).
 - Undirected loopy graph case uses a modified version of the previous simple graph case implementation (Király). It could be sped up with a similar approach to the simple graph case, but I simply don't have the time to do this now. 
 - Realizing degree sequences is now possible also for loopless and loopy undirected multigraphs, with a nice analogue of the three Havel-Hakimi variants we used for the simple graph case. It even works out that `IGRAPH_REALIZE_DEGSEQ_SMALLEST` guarantees that we get a connected graph. *Why* this works is not entirely trivial, if you want the proof just email me.
 - I did not implement graph realization for the loopy simple graph case yet. There is a simple way to do it (make a loop on every vertex with degree >=2 before we start), but it's not a nice analogue of the methods used for the other cases. I want something nicer, but don't have the time to fuss with the math right now.  Same goes for directed cases. The directed case is easier and I'll probably do it later this summer, but not in this PR. If everything were implemented, we'd have `2*2*2*3 = 24` cases (!) with the three kinds of methods ...
 - Some bugs were fixed in realizing degree sequences: sometimes negative degrees were not detected.
 - There's some ugliness in how some tests are done. I used macros instead of functions. Functions would be better, but macros work, and I don't have the energy to make this prettier ...

szhorvat commented:
@ntamas A careful review would be appreciated, but please don't merge it, just let me know if you find any issues, or if anything is unclear in the docs. This was a lot more work than I expected ...

I'll have another read-through in a couple of days and I'll merge it myself (after you approved it).

szhorvat commented:
@ntamas I added some more tests today which revealed some (old) issues. I will commit the fix to those directly to `master` as well.  It's good now, ready for review. AppVeyor fails again, but that's not because of this PR.

ntamas commented:
AppVeyor failures are caused by [this issue](https://help.appveyor.com/discussions/problems/27791-msys2-keyring-on-vs2017-image-at-least-needs-update). I'd rather wait a few more days and see if they fix it on AppVeyor's side rather than introducing a fix on our side (i.e. importing the missing PGP key manually) because it is likely to affect all projects using MSYS2 so chances are that something will be done in AppVeyor anyway.

ntamas commented:
Also, I'll start reviewing it tomorrow, thanks for your work on this!

szhorvat commented:
What I like to do when reviewing new functionality is to integrate it into the Mathematica interface.

It would be really nice if this (especially [realize_degree_sequence](https://github.com/igraph/python-igraph/issues/316)) could be integrated into the Python interface. Then I could also reference it in my paper, where I show that the "smallest" method always creates a connected graph.

BTW there's no hurry at all in reviewing this. It is not blocking any of my current projects.

ntamas commented:
Seems okay to me in general - only a few minor changes requested.

szhorvat commented on src/graphicality.c:
```diff
@@ -0,0 +1,936 @@
+ * known sufficient and necessary conditions for a degree sequence to be
+ * valid.
+ *
+ * \deprecated-by igraph_is_graphical 0.9
```
Absolutely! But I noticed that you placed it last in other functions (unless I misread it). That's why I did this. I'll move it to the top.

szhorvat commented on src/graphicality.c:
```diff
@@ -0,0 +1,936 @@
+ * the check also when either self-loops, multi-edge, or both are allowed in the graph.
+ *
+ * </para><para>
+ * For simple undirected graphs, the Erdős-Gallai conditions are checked using the linear-time
```
It passes, but I didn't check if UTF-8 is *actually supported* or just happens to work accidentally ...

szhorvat commented:
@ntamas One more thing: When I fixed the Erdos-Gallai test on `master`, I added the `igraph_vector_reverse_sort` function, and also added it to the documentation. If there is no objection, I am going to remove it from the docs on `master` and keep it on `develop` only. There's some way til 0.9 and 0.8.3 is not the right time to officially introduce new functions. I'll keep the function (since it's actually needed for the fix), but I won't include it in the docs.

szhorvat commented on src/graphicality.c:
```diff
@@ -0,0 +1,936 @@
+            *res = 0;
+            return IGRAPH_SUCCESS;
+        }
+        sum += d;
```
Fixed. I opted to fix it only here, but not in the Erdos-Gallai test as there we check that `dmax <= n-1`, so problems from overflow are exceedingly unlikely.

Actually, if I think about it: parity should be maintained even if there is integer overflow.

szhorvat commented:
I addressed everything you pointed out. I'll merge this myself in a couple of days.
